Recorded Runs

bodom0015 opened the pull request:
## Problem
We need a way for a user to view and execute Recorded Runs using the WholeTale UI.

Starts to fix #122
Fixes #123  
Fixes #124 
Fixes #204 

## Approach

* #124: Added viewer to Run > Files for Tale Versions / Recorded Runs
<img width="1680" alt="Screen Shot 2021-09-13 at 4 42 06 PM" src="https://user-images.githubusercontent.com/1413653/133167476-27dc3d8f-d590-4190-9958-5a97b412923e.png">
<img width="1679" alt="Screen Shot 2021-09-13 at 11 18 54 AM" src="https://user-images.githubusercontent.com/1413653/133120818-21563ea3-4a40-4cdd-b91b-abee79a268ad.png">

* #204: Existing Remove/rename will not work, as it uses the `/api/v1/folder` endpoints. Replaced (for this case only) with the `/api/v1/run` and `/api/v1/version` endpoints where appropriate

* #123: Added ability to launch recorded runs and view their job status
    * Uncomment and wire up the "Recorded Run" buttons (NOTE: the dropdown here still does not reflect reality)

<img width="677" alt="Screen Shot 2021-09-13 at 11 21 47 AM" src="https://user-images.githubusercontent.com/1413653/133120785-eacef7ea-16b3-4a29-b291-52d1e331eb8c.png">
<img width="1679" alt="Screen Shot 2021-09-13 at 12 15 07 PM" src="https://user-images.githubusercontent.com/1413653/133127952-0487497a-e61e-4e09-9588-99f3defec968.png">


* [WIP] Uncomment and wire up the "Edit Configurations" button (NOTE: the dropdown here still does not reflect reality)
* [WIP] Add EditRunConfigsDialogComponent, a modal to create/remove/edit run configurations
* Commented this out / hid it again - needs API/design and can be included in a future release if necessary

<img width="380" alt="Screen Shot 2021-06-22 at 5 13 40 PM" src="https://user-images.githubusercontent.com/1413653/123006626-9b037580-d37d-11eb-9f96-e698eb393665.png">
<img width="442" alt="Screen Shot 2021-06-22 at 5 13 24 PM" src="https://user-images.githubusercontent.com/1413653/123006628-9b9c0c00-d37d-11eb-90e2-f8ebcac1c2d3.png">

NOTE: This will need further styling/touch-ups and wiring (once the API is complete), but should serve as a jumping-off point for some of the discussions surrounding this feature

## How to Test
Prerequisites: at least one Tale created (does not need to be running)

### View / Rename / Delete Tale Versions
1. Navigate to the Run > Files view for a Tale that you own
    * You should see new navs on the left for `Tale Versions` and `Recorded Runs`
2. Save a new version of the Tale, then click the "Tale Versions" nav
    * You should see the version that you created listed here, as well as in the Tale Versions panel
3. On the Run > Files view, expand the dropdown beside a version and choose Rename
    * You should be prompted for a new name
4. Enter a new name and submit the form
    * You should see that the view is automatically updated to reflect the new name
5. On the Run > Files view, expand the dropdown beside a version and choose Remove
    * You should see that the view is automatically updated to delete the chosen version from the list

### Start Recorded Run
1. Checkout this branch locally, rebuild the dashboard
2. Login to the WholeTale Dashboard
3. Navigate to the Run view for a Tale that you own
4. Expand the Tale Versions panel on the right side (or the Tale dropdown at the top-right)
    * You should see a new button here to "Perform Recorded Run"
5. Click the button to "Perform Recorded Run"
    * You should see a prompt appear asking for the name
    * NOTE: This prompt needs better styling
6. Enter a name and submit the prompt
    * You should see the Event Viewer come up
    * NOTE: this prompt will eventually go away, but is a simple workaround for now
7. Click "View Logs"
    * You should see the job log output from the recorded run

### View / Rename / Delete Recorded Runs
1. Navigate to the Run > Files view for a Tale that you own
    * You should see new navs on the left for `Tale Versions` and `Recorded Runs`
2. Click on the Recorded Runs nav
    * You should see the Recorded Run listed here that you created previously
3. On the Run > Files view, expand the dropdown beside a run and choose Rename
    * You should be prompted for a new name
4. Enter a new name and submit the form
    * You should see that the view is automatically updated to reflect the new name
5. On the Run > Files view, expand the dropdown beside a run and choose Remove
    * You should see that the view is automatically updated to delete the chosen run from the list

### [HIDDEN] Run Configs
1. Navigate to the Run view for a Tale that you own
2. Expand the Tale Versions panel on the right side
    * You should see a new button here to "Edit Run Configurations"
3. Click the "Edit Run Configuration" button
    * The "Run Configurations" modal should appear (aka `EditRunConfigsDialogComponent`)
    * TODO: Wire this up to `GET /run/:id/config` or similar
4. Enter a value for `name` and `mainEntrypoint` hit Apply
    * You should see the modal close
    * TODO: Wire this up to `PUT /run/:id/config` or similar
5. Click the "Edit Run Configuration" button again
    * The "Run Configurations" modal should appear again
    * You should see the value you previously saved is still here (until UI is refreshed)
<img width="1679" alt="Screen Shot 2021-09-13 at 12 15 07 PM" src="https://user-images.githubusercontent.com/1413653/133127942-d23a33c0-ee35-4061-be56-ddac8eb162aa.png">

bodom0015 commented:
Simplified modal / test case for now... we can always bring these UI elements back if desired.

We should probably start discussing/finalizing the API spec for this stuff as well

craig-willis commented:
A few more issues noted during testing:
* Back arrow button results in console error (`Cannot read properties of null (reading `_id`)`)
* When deleting a version that has an associated run, the 461 error is swallowed. The user can't delete versions that have runs.
* Runs do not (yet) appear in the Tale history as per the mockups (https://whole-tale.github.io/wholetale-css-mockup/src/run.html). You may have mentioned this during a call and it could be a separate PR.
* I noticed that runs weren't appearing in the container. This is being addressed in https://github.com/whole-tale/girderfs/pull/29, https://github.com/whole-tale/wt_versioning/pull/34, https://github.com/whole-tale/gwvolman/pull/155. 

Finally, I'm proposing that we temporarily do away with the run config modal and simply add the entrypoint script name to the `run/{id}/start` endpoint.  I think we can revisit the run config after we get user feedback. Comments welcome here https://github.com/whole-tale/gwvolman/pull/156.

bodom0015 commented:
I believe I've addressed the issues pointed out above:
* Handle 461 error by showing error modal
* Handle 303 error when creating a version by grabbing latest version and using that instead
* Fixed an error with Run > Files > Runs back button
* Removed "run name" prompt, added a form field to start a run using a particular entrypoint

Also fixed:
* Enabled strict type checking in templates - this should provide better/more detailed error messages when template errors occur
* An error where clicking the "External Data" nav would continue to show the previous nav's file data

bodom0015 commented:
Didn't notice that the 303 was returning the `extra` id in the error body (for some reason, I was expecting this in the headers)...

I've updated the "Perform Recorded Run" code to utilize this `extra` field instead of manually looking up the most recent version :+1:

bodom0015 commented:
I may be forgetting some of the outcomes here, but I feel like these were issues that were brought up at the dev meeting...

* I recall mentioning this last week.. there is definitely a disconnect between the dropdown and the panel options, but I'm also unclear why we need the dropdown item at all anymore? I could just add another modal here with a single text input, but I feel like we have several of those already for one-offs like this and wanted to avoid making another one unless it's really needed 
* Delete/Rename Run offered on Run > Files > Recorded Runs, but not on the history panel. I'm not sure how much of the functionality that is now available on Run > Files should be duplicated on the Tale Versions panel, but I've opted not to duplicate things where possible

My feeling, given the above, is that we now have too many ways to access Versions/Runs... Is there really a need for two of every button for this new feature? Perhaps we can get rid of the dropdown menu item for "Perform Recorded Run" altogether? Every function available of the Tale Versions panel is also available elsewhere (with the exception of this new `entrypoint` input)

Xarthisius reviewed (approved):
I think it reached a state when we should merge and file new issues and necessary. I tested basic features and it looks good to me. Couple of minor issues:

1. Created a Tale, Tale > Files > (ellipsis) > Recorded Run:
   * Title of the modal says: "Connect to Git Repository"
1. Dropdown for in Tale > Files > Versions > <some_version> > "v" disappears almost immediately after click, which makes e.g renaming version close to impossible for someone with boomer APM like me.

bodom0015 commented:
Thanks, @Xarthisius! Good catch there.. I've updated the modal title for now to instead say `Recorded Run`

I've witnessed the dropdown behavior as well and can agree it is very annoying. I'm not sure why the menu seems to close so quickly, but my current suspicion is that Angular is constantly re-rendering the view and that re-render forces the menu to close. Still unclear on what part of the `ngOnChanges` there needs to be updated, but can file this as a new issue to make sure it is addressed in the near future :+1:

craig-willis reviewed (approved):
I agree with @Xarthisius that we should merge this and address anything further in separate issues. Noting a few things here that I encountered during testing
* Same issue with dropdowns reported above
* Renaming a version changes order in history panel (should be based on created date instead?)
* Renamed version is not reflected in "Run based on <old name>"
* Link to version from "Run based on..." is nice, can we do the same for the version and run names (i.e., the names link to the version/run in the file browser)?

bodom0015 commented:
:+1: Filed new issues to address these items:
* #234 
* #235 

Will merge this up and work toward addressing the remaining issues
